src/models/transactions.ts

```diff
@@ -0,0 +1,163 @@
+import {
+  connection,
+  Schema,
+  model,
+  Model,
+  Types,
+  PreSaveMiddlewareFunction,
+  Document,
+} from "mongoose";
+import { extractObjectIdKeysFromSchema } from "../helper/extractObjectIdKeysFromSchema";
+import type { IAnyObject } from "../IAnyObject";
+import type { IJournal } from "./journals";
+import { lockModel } from "./lock";
+
+export interface ITransaction {
+  _id: Types.ObjectId;
+  credit: number;
+  debit: number;
+  meta: IAnyObject;
+  datetime: Date;
+  account_path: string[];
+  accounts: string;
+  book: string;
+  memo: string;
+  _journal: Types.ObjectId | IJournal;
+  timestamp: Date;
+  voided: boolean;
+  void_reason?: string;
+  approved: boolean;
+  _original_journal?: Types.ObjectId | IJournal;
+}
+
+export const transactionSchema = new Schema<ITransaction>(
+  {
+    credit: Number,
+    debit: Number,
+    meta: Schema.Types.Mixed,
+    datetime: Date,
+    account_path: [String],
+    accounts: String,
+    book: String,
+    memo: String,
+    _journal: {
+      type: Schema.Types.ObjectId,
+      ref: "Medici_Journal",
+    },
+    timestamp: Date,
+    voided: {
+      type: Boolean,
+      default: false,
+    },
+    void_reason: String,
+    // The journal that this is voiding, if any
+    _original_journal: {
+      type: Schema.Types.ObjectId,
+      ref: "Medici_Journal",
+    },
+    approved: {
+      type: Boolean,
+      default: true,
+    },
+  },
+  { id: false, versionKey: false, timestamps: false }
+);
+
+export let transactionModel: Model<ITransaction>;
+
+let transactionSchemaKeys: Set<string> = new Set(
+  Object.keys(transactionSchema.paths)
+);
+
+const preSave: PreSaveMiddlewareFunction<ITransaction & Document> =
+  async function (next) {
+    if (
+      !this.isModified("approved") ||
+      // @ts-ignore
+      (this.$isNew === true && this.approved === false) ||
+      !this.$session()
+    ) {
+      return next();
+    }
+
+    if (
+      this.$locals.lock &&
+      (this.$locals.lock as string[]).indexOf(this.accounts) === -1
+    ) {
+      return next();
+    }
+
+    const session = this.$session();
+
+    const book = this.book;
+    const account = this.accounts;
+
+    await lockModel.collection.updateOne(
+      { account, book },
+      { $inc: { __v: 1 } },
```

Comment:
Why does this work? Should be also containing $setOnInsert: {book, account}

Not that mongo just creates a general document and blocks everything